lib/rust/profiling/src/macros.rs

```diff
@@ -0,0 +1,204 @@
+#[allow(missing_docs)]
+#[macro_export]
+macro_rules! block_wrapper_start {
+    (false, $fn_start:ident, $fn_end:ident, [$($expr:stmt;)*]) => {
+        $crate::make_call!{$fn_start}
+        $crate::block_wrapper_continue!{$fn_end, $($expr;)*}
+    };
+    (true, $fn_start:ident, $fn_end:ident, [$($expr:stmt;)*]) => {
+        $crate::make_call!{$fn_start}
+        $($expr;)*
+        $crate::make_call!{$fn_end}
+    };
+}
+#[allow(missing_docs)]
+#[macro_export]
+macro_rules! block_wrapper_continue {
+    ($fn_end:ident, $expr:stmt;) => {
+        let out = { $expr };
+        $crate::make_call!{$fn_end}
+        out
+    };
+    ($fn_end:ident, $expr:stmt; $($exprs:stmt;)*) => {
+        $expr;
+        $crate::block_wrapper_continue!{$fn_end:ident, $($exprs;)*}
+    };
+}
+
+/// Create a Metadata struct that has the source location prepopulated via the `file!` and
+/// `line!` macros.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! make_metadata {
+    ($profiling_level:expr, $interval_name:expr) => {
+        $crate::Metadata {
+            source:          $crate::SourceLocation { file: file!().to_string(), line: line!() },
+            profiling_level: $profiling_level.to_string(),
+            label:           $interval_name.to_string(),
+        }
+    };
+}
+
+/// Start measuring the interval of the given profiling level and name.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! start_interval {
+    ($profiling_level:expr, $interval_name:expr) => {
+        $crate::mark_start_interval($crate::make_metadata!($profiling_level, $interval_name))
+    };
+}
+
+/// End measuring the interval of the given profiling level and name.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! end_interval {
+    ($profiling_level:expr, $interval_name:expr) => {
+        $crate::warn_on_error($crate::mark_end_interval($crate::make_metadata!($profiling_level, $interval_name)))
+    };
+}
+
+/// Measure an interval with the given profiling level and name that starts before the given
+/// expressions and ends after the given expressions.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! measure_interval {
+        ($d:tt, $profiling_level:expr, $interval_name:expr, $($body:tt)*) => {
+             {
+
+                 fn start() {
+                     $crate::start_interval!($profiling_level,$interval_name).release();
+                 }
+
+                 fn end() {
+                     $crate::end_interval!($profiling_level,$interval_name);
+                 }
+
+                 $crate::wrap_block! { start end $($body)* }
+             }
+        }
+    }
+
+
+
+/// Define a new boolean variable whose value is determined by a feature flag in the crate.
+/// The name of the variable is `ENABLED` and it will be true if a feature flag
+/// `enable-<profiling_level_name>-profiling` is set, otherwise it will be false.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! define_profiling_toggle {
+    ($profiling_level_name:ident) => {
+        paste::item! {
```

Comment:
why you need paste::item for? I dont see any usage of it and it breaks IntelliJ code reasoning mechanisms.

Reply:
It is needed to allow the `#[cfg(feature = "enable-" $profiling_level_name "-profiling")]` syntax required to enable conditional compilation. Also, for the doc string.